File: quizzes2/07-problem.ts
export {};

// @ts-ignore
import * as fs from 'fs';

type Feedback = {
    comment: string;
    date: string;
};

type Result = {
    rating: number;
    feedback: Feedback[];
};

type Staff = {
    id: number;
    full_name: string;
    email: string;
    result: Result;
};

type LatestFeedback = Pick<Staff, 'email' | 'full_name'> & {
    latest_comment: string;
};

const staffs: Staff[] = JSON.parse(fs.readFileSync('input_data/staff.json'));

const findLatestFeedbacksWithChained = (): LatestFeedback[] => {
    return staffs.map(({ result, ...staff }) => {
        const latest: Feedback = result.feedback.toSorted(
            (a, b) => new Date(b.date).getTime() - new Date(a.date).getTime(),
        )[0];

        return {
            ...staff,
            latest_comment: latest?.comment ?? 'No available feedback yet.',
        };
    });
};

const findLatestFeedbacksWithLoop = (): LatestFeedback[] => {
    const latestFeedbacks: LatestFeedback[] = [];

    for (const { result, ...staff } of staffs) {
        const feedbacks: Feedback[] = result.feedback;

        let latest: Feedback | null = null;

        for (const feedback of feedbacks) {
            if (!latest || new Date(feedback.date) > new Date(latest.date)) {
                latest = feedback;
            }
        }

        latestFeedbacks.push({
            ...staff,
            latest_comment: latest?.comment ?? 'No available feedback yet.',
        });
    }

    return latestFeedbacks;
};

console.log(
    JSON.stringify({ data: findLatestFeedbacksWithChained() }, null, 2),
);
console.log(JSON.stringify({ data: findLatestFeedbacksWithLoop() }, null, 2));

File: bins5/07-problem.rb

# frozen_string_literal: true

require 'json'
require 'date'

staffs = JSON.parse(
  File.read('input_data/staff.json'),
  symbolize_names: true
)

def find_latest_feedbacks_with_chain(staffs)
  staffs.map do |staff|
    result = staff[:result]
    latest = result[:feedback].max_by do |f|
      Date.parse(f[:date])
    end

    staff.slice(:email, :full_name).merge(latest_comment: latest&.dig(:comment) || 'No available feedback yet.')
  end
end

def find_latest_feedbacks_with_loop(staffs)
  latest_feedbacks = []

  staffs.each do |staff|
    latest = nil

    staff[:result][:feedback].each do |f|
      if latest.nil? || Date.parse(f[:date]) > Date.parse(latest[:date])
        latest = f
      end
    end

    latest_feedbacks << staff.slice(:email, :full_name).merge(latest_comment: latest&.dig(:comment) || 'No available feedback yet.')
  end

  latest_feedbacks
end

puts JSON.pretty_generate({ data: find_latest_feedbacks_with_chain(staffs) })
puts JSON.pretty_generate({ data: find_latest_feedbacks_with_loop(staffs) })
